### PR TITLE
Start adding an `aarch64-pc-windows-msvc` target

### PR DESCRIPTION
**Pull request by alexcrichton**
This commit adds the necessary definitions for target specs and such as well as
the necessary support in libstd to compile basic `aarch64-pc-windows-msvc`
binaries. The target is not currently built on CI, but it can be built locally
with:

    ./configure --target=aarch64-pc-windows-msvc --set rust.lld
    ./x.py build src/libstd --target aarch64-pc-windows-msvc

Currently this fails to build `libtest` due to a linker bug (seemingly in LLD?)
which hasn't been investigate yet. Otherwise though with libstd you can build a
hello world program (linked with LLD). I've not tried to execute it yet, but it
at least links!

Full support for this target is still a long road ahead, but this is hopefully a
good stepping stone to get started.

Points of note about this target are:

* Currently defaults to `panic=abort` as support is still landing in LLVM for
  SEH on AArch64.
* Currently defaults to LLD as a linker as I was able to get farther with it
  than I was with `link.exe`

### CODE REVIEW

**Comment by upsuper** [comment]
Given [the test file in LLVM repo](https://github.com/llvm-mirror/llvm/blob/36f54002c931a026f490f9fb074c11d91e3487a2/test/CodeGen/AArch64/dllimport.ll#L1-L3), shouldn't this call `aarch64-unknown-windows-msvc` instead?

**Comment by alexcrichton** [comment]
Perhaps! I'm thinking though it's probably best to stick to the convention we already have for our MSVC targets

**Comment by TimNN** [comment]
Ping from triage @sfackler. This PR requires your review.

**Comment by retep998** [comment]
cc https://github.com/rust-lang/rust/issues/52659

File: src/libpanic_unwind/macros.rs [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+
+/// A macro for defining `#[cfg]` if-else statements.
+///
+/// This is similar to the `if/elif` C preprocessor macro by allowing definition
+/// of a cascade of `#[cfg]` cases, emitting the implementation which matches
+/// first.
+///
+/// This allows you to conveniently provide a long list `#[cfg]`'d blocks of code
+/// without having to rewrite each clause multiple times.
+macro_rules! cfg_if {
```
Comment (sfackler) [inline src/libpanic_unwind/macros.rs]: Don't we already pull the cfg_if crate in?

File: src/librustc_target/spec/aarch64_pc_windows_msvc.rs [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+use spec::{LinkerFlavor, Target, TargetResult, PanicStrategy, LldFlavor};
+
+pub fn target() -> TargetResult {
+    let mut base = super::windows_msvc_base::opts();
+    base.max_atomic_width = Some(64);
+    base.has_elf_tls = true;
+
+    // FIXME: this shouldn't be panic=abort, it should be panic=unwind
+    base.panic_strategy = PanicStrategy::Abort;
+    base.linker = Some("rust-lld".to_owned());
```
Comment (sfackler) [inline src/librustc_target/spec/aarch64_pc_windows_msvc.rs]: Is the use of LLD a fixme or the right long term approach?

**Pull request merged**